### PR TITLE
Fix FEATHER_S2 update

### PR DESCRIPTION
**Pull request by MichaelWheeley**
## Description
- For FeatherS2 only, after querying chipID create a 15s delay.
- For FeatherS2 only, after flash erase create a 15s delay.
- minor spelling update to user interface developement -> development.

## Motivation and Context
Adafruit FeatherS2 appears not to comply with 'no_reset' command in ESP programming tool
for instance in the RunEspTool() Process call the arguments for .StartInfo are,
For ChipID check: "--port COM6  --chip auto   --after no_reset flash_id"
For flash erase "--port COM6  --chip esp32s2   --after no_reset erase_flash"
however after each operation the device appears to reset and a USB enumeration occurs. There is no existing time allowance in the code for this and subsequent operation occurs when no COM port exists resulting in 100% fail rate.

## How Has This Been Tested?
Code has been tested against Adafruit FeatherS2. It is unknown what other devices may exhibit this issue.

## Screenshots
without fix the following exception occurs,
image

## Types of changes
- [X] Improvement (non-breaking change that improves a feature, code or algorithm)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nanoFirmwareFlasher?pullRequest=92) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nanoFirmwareFlasher?pullRequest=92) before we can accept your contribution.<br/><br/>:x: MichaelWheeley [sign now](https://cla.dotnetfoundation.org/nanoframework/nanoFirmwareFlasher?pullRequest=92)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nanoFirmwareFlasher?pullRequest=92) it.</sub>

File: nanoFirmwareFlasher/Esp32Operations.cs [hunk 1]
```diff
@@ -284,6 +287,13 @@ internal class Esp32Operations
 
             if (updateFw)
             {
+#if FEATHER_S2_DELAY
+                if (targetName.Contains("FEATHER_S2"))
+                {
+                    Thread.Sleep(TimeSpan.FromSeconds(15));
```
Comment (networkfusion) [inline nanoFirmwareFlasher/Esp32Operations.cs:293]: Might be worth adding an explanation as to why 15 seconds is optimal

File: nanoFirmwareFlasher/Esp32Operations.cs [hunk 2]
```diff
@@ -303,6 +313,13 @@ internal class Esp32Operations
 
             if (operationResult == ExitCodes.OK)
             {
+#if FEATHER_S2_DELAY
+                if (targetName.Contains("FEATHER_S2"))
+                {
+                    Thread.Sleep(TimeSpan.FromSeconds(15));
```
Comment (networkfusion) [inline nanoFirmwareFlasher/Esp32Operations.cs:319]: Might be worth adding an explanation as to why 15 seconds is optimal

File: nanoFirmwareFlasher/Esp32Operations.cs (hunk 1 above)
Comment (MichaelWheeley) [reply nanoFirmwareFlasher/Esp32Operations.cs:293]: tried 5s and it was unreliable, noticed it fail occasionally at 10s, 15s seemed to be the minimum for guaranteed success

File: nanoFirmwareFlasher/Esp32Operations.cs (hunk 2 above)
Comment (MichaelWheeley) [reply nanoFirmwareFlasher/Esp32Operations.cs:319]: tried 5s and it was unreliable, noticed it fail occasionally at 10s, 15s seemed to be the minimum for guaranteed success

File: nanoFirmwareFlasher/Esp32Operations.cs (hunk 1 above)
Comment (networkfusion) [reply nanoFirmwareFlasher/Esp32Operations.cs:293]: Yes, I understand (saw the discord comment), but that will be lost, so better to make it a code comment 😉

**Comment by MichaelWheeley** [comment]
perhaps there is an under laying reason the device resets after each operation and/or does not recognize the --after no_reset parameter. This could be a short term fix if the under laying reason is resolved.

File: nanoFirmwareFlasher/Esp32Operations.cs (hunk 1 above)
Comment (MichaelWheeley) [reply nanoFirmwareFlasher/Esp32Operations.cs:294]: done! :-)

File: nanoFirmwareFlasher/Esp32Operations.cs (hunk 2 above)
Comment (MichaelWheeley) [reply nanoFirmwareFlasher/Esp32Operations.cs:321]: done! :-)

**Comment by josesimoes** [comment]
@MichaelWheeley thank you very much for looking into this! 😃 

I've done a quick test here, and I now understand the need for such a "long" waiting time. It occurs if you happen to have VS open with nanoFramework extension installed. In this situation, the auto-detect feature will be trying to connect to that "new" COM port that has "just" arrived to the system after that reenumeration after the flash erase.

Can you please double check that at your end? Just shutdown VS and it should work without needing any timeout.

To your point: what we need to address is the fact that the device it's internally resetting after the flash erase operation. It shouldn't. As soon as we can "fix" that all the rest will go away.

**Comment by MichaelWheeley** [comment]
> @MichaelWheeley thank you very much for looking into this! 😃
> 
> I've done a quick test here, and I now understand the need for such a "long" waiting time. It occurs if you happen to have VS open with nanoFramework extension installed. In this situation, the auto-detect feature will be trying to connect to that "new" COM port that has "just" arrived to the system after that reenumeration after the flash erase.
> 
> Can you please double check that at your end? Just shutdown VS and it should work without needing any timeout.
> 
> To your point: what we need to address is the fact that the device it's internally resetting after the flash erase operation. It shouldn't. As soon as we can "fix" that all the rest will go away.

**Comment by MichaelWheeley** [comment]
@josesimoes yes that appears to be correct, I reduced the delay to 1s in both cases, recompiled, close VS and ran nanoff from a separate PowerShell window. Worked fine though for safety 1s may need to be 3s.
Then tried removing the delays altogether, it goes through the ChipID/Erase transition with no problem, but fails 100% at the Erase/Program transition unless there is a delay.

**Comment by josesimoes** [comment]
> @josesimoes yes that appears to be correct, I reduced the delay to 1s in both cases, recompiled, close VS and ran nanoff from a separate PowerShell window. Worked fine though for safety 1s may need to be 3s.
> Then tried removing the delays altogether, it goes through the ChipID/Erase transition with no problem, but fails 100% at the Erase/Program transition unless there is a delay.
> 
> 

OK that makes more sense and coherent with hardware behavior!
Agreed on the 3s VS 1s.

Did you checked the parameters being passed to esptool on the ID stage and flash erase? I was under the impression that I had it figured out and got rid of that unwanted restart... Maybe at some point I've messed it up with so many trials and attempts. Perhaps you can spot something that I left there wrongly or maybe forgotten.

**Comment by MichaelWheeley** [comment]
my notes only,
`esptool /?
usage: esptool [-h] [--chip {auto,esp8266,esp32,esp32s2,esp32s3beta2,esp32s3beta3,esp32c3,esp32c6beta}] [--port PORT]
               [--baud BAUD] [--before {default_reset,usb_reset,no_reset,no_reset_no_sync}]
               [--after {hard_reset,soft_reset,no_reset,no_reset_stub}] [--no-stub] [--trace]
               [--override-vddsdio [{1.8V,1.9V,OFF}]] [--connect-attempts CONNECT_ATTEMPTS]
               {load_ram,dump_mem,read_mem,write_mem,write_flash,run,image_info,make_image,elf2image,read_mac,chip_id,flash_id,read_flash_status,write_flash_status,read_flash,verify_flash,erase_flash,erase_region,merge_bin,version,get_security_info}`

**Comment by MichaelWheeley** [comment]
interesting, if you replace no_reset with no_reset_stub then error message is suppressed and no reset occurs, exactly the characteristic intended... so then it works and the delay can be removed entirely.
However not tested in any other device/environment so leave it to you @josesimoes to recommend next step

btw, private bool _isStubActive is set but not used, unnecessary and confusing so removed it

**Comment by josesimoes** [comment]
> interesting, if you replace no_reset with no_reset_stub then error message is suppressed and no reset occurs, exactly the characteristic intended... so then it works and the delay can be removed entirely.
> However not tested in any other device/environment so leave it to you @josesimoes to recommend next step

Awesome!! That's what I was seeing at some point!! 😅
Now we should test these with the other series and also this on but without download mode. 


> 
> btw, private bool _isStubActive is set but not used, unnecessary and confusing so removed it
> 

Correct. I remove the code that was making use of it but forgot to remove this one. Nice catch.

**Review by josesimoes — APPROVED** [review]
Tested with several ESP32 targets (and also with ESP32_S2, all updated fine so... we have a winner!
Thanks for tackling this one. 💯 👏🏻

**Pull request merged**